wicket-core/src/main/java/org/apache/wicket/markup/ComponentTag.java

```diff
@@ -126,6 +127,8 @@ public class ComponentTag extends MarkupElement
 
 	private IAutoComponentFactory autoComponentFactory;
 
+	private static final Map<String,AppendingStringBuffer> tagMap = new ConcurrentHashMap<>();
```

Comment:
I feel a bit uneasy about using an unbounded map. Since Wicket might be used to parse arbitrary XML, we have no way of knowing how many distinct tags users may have in their application. This will probably be fine for 99.9% of use cases, but it might cause issues for some users.

Possible alternatives: 

- Create a cache map of all known HTML5 tags
- Use a bounded map. We currently only seem to have a `MostRecentlyUsedMap`, that is not thread-safe.

Does anyone else have better ideas?

I'm very interested in this optimization because GC pressure under high load is the main problem in our application.

Reply:
Why not go for a much easier approach: keep the method as is and mark it deprecated in wicket 9.x and add a new method `writeSyntheticCloseTag(Response)` that writes the close tag directly to the `Response`. That would remove all the object allocations and simplify the code a lot. The solution with a cache seems overly complicated and might introduce congestion with lookups and updates.

Reply:
@theigl I appreciate your concern about an unbounded map.   I toyed with the idea of pre-populating the map with a small set of frequently used tags.  That's a small change and gets 99% of the benefit.   I wrote the change this way as it was trivial and gets the conversation started.

I will tweak the code slightly in the case when a buffer is still allocated to size the internal character array properly and to be less wasteful.


@papegaaij Note that the original allocating code runs in twice the time of the map lookups, and has a greater probability of contention caused by GC pauses than any possible contention in the map (ConcurrentHashMap only contends on structural updates, which will be very rare after the few lines of HTML are rendered in the application).

Reply:
@astange1 the original code is not very optimal indeed, but I don't see why this method needs to return a closing tag, when it's only called from 2 sites, and both sites immediately append the result to a response. Simply skipping the whole return value and appending the closing tag to the response avoids any allocations, does not need a global synchronized store, avoids a hash lookup and is a lot easier to understand. Even though this method is strictly speaking part of the public API of wicket, it is not meant to be used externally. Changing the method to better suit the needs is ok, as long as we keep the old method for strict API compatibility in 9.x. We can drop it in 10.

Reply:
Excellent.   I figured there must have been a better way of doing this.